File: src/prepare_data_for_gbm.r
require(dplyr)
require(stringr)
require(aeneval)

prepare_data <- function(
  index,
  candidates,
  label_disribution,
  include_ground_truth = FALSE,
  ground_truth = NULL,
  max_docs = -1
) {
  if (max_docs > 0) {
    message("Using only ", max_docs, " documents.")
    index <- index  |>
      slice_head(n = max_docs)
  }

  if (include_ground_truth) {

    gold_standard <- index %>%
      inner_join(gt, by = c("idn", "kind"), na_matches = "never")  |>
      mutate(
        doc_id = idn,
        label_id = str_match(uri, "(?<idn>[0-9X]{9,10})")[, "idn"]
      )

    n_idn_not_predicted <- anti_join(
      distinct(gold_standard, doc_id),
      distinct(candidates, doc_id),
      by = "doc_id"
    )  |> nrow()

    if (n_idn_not_predicted > 0) {
      message(
        "There are ",
        n_idn_not_predicted,
        " documents in the ground truth with no candidates."
      )
    }

    res <- create_comparison(
      gold_standard,
      candidates
    )  |>
      left_join(label_disribution, by = "label_id", na_matches = "never")

    res <- res  |>
      filter(suggested)  |>
      mutate(gold = ifelse(gold, 1.0, 0.0))  |>
      select(
        gold,
        score,
        label_freq,
        occurrences,
        first_occurence,
        last_occurence,
        spread,
        is_prefLabel
      )
  } else {
    index <- select(index, doc_id = idn)

    n_idn_not_predicted <- anti_join(
      index,
      distinct(candidates, doc_id),
      by = "doc_id"
    )  |> nrow()

    if (n_idn_not_predicted > 0) {
      message(
        "There are ",
        n_idn_not_predicted,
        " documents in the index with no candidates."
      )
    }

    res <- index  |>
      inner_join(candidates, by = "doc_id", na_matches = "never")  |>
      left_join(label_disribution, by = "label_id", na_matches = "never")  |>
      select(
        doc_id,
        label_id,
        score,
        label_freq,
        occurrences,
        first_occurence,
        last_occurence,
        spread,
        is_prefLabel
      )
  }
}

File: src/prepare_data_for_gbm.py
import pandas as pd

pd.set_option("future.no_silent_downcasting", True)


def create_comparison(
    gold_standard: pd.DataFrame, predictions: pd.DataFrame
) -> pd.DataFrame:
    outer = gold_standard.merge(predictions, on=["doc_id"], how="outer", indicator=True)
    gold_wo_predicted = outer[outer._merge == "left_only"].drop("_merge", axis=1)
    predicted_wo_gold = outer[outer._merge == "right_only"].drop("_merge", axis=1)

    if len(predicted_wo_gold.index) != 0:
        return
    if len(gold_wo_predicted.index) > 0:
        print("Gold standard data contains documents that are not in predicted set")

    gold_standard["gold"] = True
    predictions["suggested"] = True
    compare = pd.merge(
        left=gold_standard,
        right=predictions,
        how="outer",
        on=["doc_id", "label_id"],
    ).fillna(value={"gold": False, "suggested": False})

    return compare


def prepare_data(
    index: pd.DataFrame,
    include_ground_truth: bool,
    ground_truth: pd.DataFrame | None,
    candidates: pd.DataFrame,
    label_distribution: pd.DataFrame,
    max_docs: int,
) -> pd.DataFrame:
    if max_docs > 0:
        print(f"Using only {max_docs} documents.")
        index = index.head(max_docs)

    if include_ground_truth:
        gold_standard = pd.merge(
            left=index, right=ground_truth, on=["idn", "kind"], how="inner"
        )
        gold_standard["doc_id"] = gold_standard["idn"]
        gold_standard["label_id"] = gold_standard["uri"].str.extract(r"([0-9X]{9,10})")

        outer = pd.merge(
            left=gold_standard[["doc_id"]],
            right=candidates[["doc_id"]],
            how="outer",
            indicator=True,
        )
        n_idn_not_predicted = len(
            outer[outer._merge == "left_only"].drop("_merge", axis=1).index
        )

        if n_idn_not_predicted > 0:
            print(
                f"There are {n_idn_not_predicted} documents in the ground truth with no candidates."
            )
        result = create_comparison(gold_standard, candidates)
        result = pd.merge(
            left=result, right=label_distribution, on=["label_id"], how="left"
        )

        result = result.loc[result["suggested"]]
        result["gold"] = [1 if i else 0 for i in result["gold"].tolist()]
        result = result.filter(
            items=[
                "gold",
                "score",
                "label_freq",
                "occurrences",
                "first_occurence",
                "last_occurence",
                "spread",
                "is_prefLabel",
            ]
        )
    else:
        index = index.filter(items=["idn"]).rename(columns={"idn": "doc_id"})

        outer = pd.merge(
            left=index, right=candidates[["doc_id"]], how="outer", indicator=True
        )
        n_idn_not_predicted = len(
            outer[outer._merge == "left_only"].drop("_merge", axis=1).index
        )

        if n_idn_not_predicted > 0:
            print(
                f"There are {n_idn_not_predicted} documents in the ground truth with no candidates."
            )

        result = pd.merge(left=index, right=candidates, on=["doc_id"], how="left")
        result = pd.merge(
            left=result, right=label_distribution, on=["label_id"], how="left"
        )
        result = result.filter(
            items=[
                "doc_id",
                "label_id",
                "score",
                "label_freq",
                "occurrences",
                "first_occurence",
                "last_occurence",
                "spread",
                "is_prefLabel",
            ]
        )

    result["is_prefLabel"] = result["is_prefLabel"].astype(bool)
    return result
